presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -81,12 +83,24 @@ public synchronized ScheduleResult schedule()
 
                 long start = System.nanoTime();
                 batchFuture = splitSource.getNextBatch(splitBatchSize);
-                batchFuture.thenRun(() -> stage.recordGetSplitTime(start));
+                Futures.addCallback(batchFuture, new FutureCallback<List<Split>>()
```

Comment:
we really need just independent onSuccess and onFailure callback handlers helpers.